File: src/main/java/com/sberbank/pfm/test/concordion/extensions/exam/db/commands/DBSetCommand.java
package com.sberbank.pfm.test.concordion.extensions.exam.db.commands;

import org.concordion.api.CommandCall;
import org.concordion.api.Element;
import org.concordion.api.Evaluator;
import org.concordion.api.ResultRecorder;
import org.dbunit.JdbcDatabaseTester;
import org.dbunit.database.IDatabaseConnection;
import org.dbunit.dataset.DefaultDataSet;
import org.dbunit.dataset.ITable;
import org.dbunit.operation.DatabaseOperation;

import static org.dbunit.operation.DatabaseOperation.CLEAN_INSERT;
import static org.dbunit.operation.DatabaseOperation.INSERT;

public class DBSetCommand extends DBCommand {
    public DBSetCommand(JdbcDatabaseTester dbTester) {
        super(dbTester);
    }

    @Override
    public void setUp(CommandCall commandCall, Evaluator evaluator, ResultRecorder resultRecorder) {
        super.setUp(commandCall, evaluator, resultRecorder);
        try {
            Element el = commandCall.getElement();
            renderTable(el, expectedTable);
            setUpDB(expectedTable, parseInsertMode(el));
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    private void setUpDB(ITable iTable, DatabaseOperation insertMode) throws Exception {
        IDatabaseConnection connection = null;
        try {
            connection = dbTester.getConnection();
            dbTester.setSetUpOperation(insertMode);
            dbTester.setDataSet(new DefaultDataSet(iTable));
            dbTester.onSetup();
        } finally {
            if (connection != null) {
                try {
                    dbTester.closeConnection(connection);
                } catch (Exception e) {
                    System.err.println("Could not close connection " + e);
                    e.printStackTrace();
                }
            }
        }
    }

    private DatabaseOperation parseInsertMode(Element el) {
        return el.getAttributeValue("mode") != null && el.getAttributeValue("mode").equals("add") ?
                INSERT : CLEAN_INSERT;
    }
}

File: exam-db/src/main/java/com/adven/concordion/extensions/exam/db/commands/DBSetCommand.kt
package com.adven.concordion.extensions.exam.db.commands

import com.adven.concordion.extensions.exam.core.html.html
import com.adven.concordion.extensions.exam.db.DbPlugin
import com.adven.concordion.extensions.exam.db.DbTester
import com.adven.concordion.extensions.exam.db.builder.ExamDataSet
import org.concordion.api.CommandCall
import org.concordion.api.Evaluator
import org.concordion.api.Fixture
import org.concordion.api.ResultRecorder
import org.dbunit.operation.DatabaseOperation
import org.dbunit.operation.DatabaseOperation.CLEAN_INSERT
import org.dbunit.operation.DatabaseOperation.DELETE
import org.dbunit.operation.DatabaseOperation.DELETE_ALL
import org.dbunit.operation.DatabaseOperation.INSERT
import org.dbunit.operation.DatabaseOperation.REFRESH
import org.dbunit.operation.DatabaseOperation.TRUNCATE_TABLE
import org.dbunit.operation.DatabaseOperation.UPDATE

class DBSetCommand(name: String, tag: String, dbTester: DbTester, pattern: DbPlugin.ValuePrinter) : DBCommand(name, tag, dbTester, pattern) {
    private val operations: Map<String, DatabaseOperation> = mapOf(
        "clean_insert" to CLEAN_INSERT,
        "insert" to INSERT,
        "update" to UPDATE,
        "refresh" to REFRESH,
        "delete" to DELETE,
        "delete_all" to DELETE_ALL,
        "truncate_table" to TRUNCATE_TABLE
    )

    override fun setUp(cmd: CommandCall?, eval: Evaluator?, resultRecorder: ResultRecorder?, fixture: Fixture) {
        super.setUp(cmd, eval, resultRecorder, fixture)
        val el = cmd.html()
        el(renderTable(el.takeAwayAttr("caption"), expectedTable, remarks, valuePrinter))
        operations.getOrElse(
            el.takeAwayAttr("operation", "clean_insert").toLowerCase(),
            { throw IllegalArgumentException("Unsupported dbunit operation. Supported: ${operations.keys}") }
        ).execute(dbTester.connectionFor(ds), ExamDataSet(expectedTable, eval!!))
    }
}
